---
title: "sc-RNAseq & sc-CHIPseq analysis of breast cancer"
author: "Alba Machado-Lopez"
date: "October 2022"
output: html_document
knit: (function(input_file, encoding) {
    out_dir <- 'docs';
    rmarkdown::render(input_file,
      encoding=encoding,
      output_file=file.path(dirname(input_file), out_dir, 'index.html'))})
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load libraries

```{r, message=F}
if (!require(R.utils)) install.packages('R.utils')
library(R.utils)

if (!require(dplyr)) install.packages('dplyr')
library(dplyr)

if (!require(stringr)) install.packages('stringr')
library(stringr)

if (!require(Seurat)) install.packages('Seurat')
library(Seurat)
```


# Download files

```{bash}
mkdir -p data
cd data
wget -r -np -nd 'https://ftp.ncbi.nlm.nih.gov/geo/series/GSE117nnn/GSE117309/suppl/GSE117309_RAW.tar' -R "index.html*" # Omitting the index file
```


# Decompress files

```{r, eval=FALSE}
# Untar the main file
untar("data/GSE117309_RAW.tar", exdir="data/")

# Untar sc-RNAseq files
RNA_files <- list.files(path="data/",pattern="scRNA")
untar_RNA_files <- function(file){untar(paste0("data/",file), exdir="data/RNA_files")}
sapply(RNA_files, untar_RNA_files)
```

# Read files into R

## sc-ChIP files
```{r}
ChIP_files <- list.files(path="data/",pattern=".txt", full.names = T)
scChIP_list <- mapply(read.table, ChIP_files)

# Clean-up object names for readability
names(scChIP_list) <- names(scChIP_list) %>% str_remove(".txt.gz") %>% gsub("data/GSM32908.._CountTable_","",.)
```

## sc-RNAseq files

```{r}
load_scRNA_files <- function(model, organism){
  directory <- paste0("data/RNA_files/filtered_gene_bc_matrices_",model,"/",organism,"/")
  expression_matrix <- ReadMtx(
  mtx=paste0(directory, "matrix.mtx"),
  features=paste0(directory, "genes.tsv"),
  cells=paste0(directory, "barcodes.tsv"))
  seurat_object <- CreateSeuratObject(counts = expression_matrix)
  return(seurat_object)
}

models <- c("HBCx-95","HBCx-95_CAPAR", "HBCx-22","HBCx22-TAMR")
organisms <- c("hg19","mm10")

scRNA_list <- mapply(load_scRNA_files, models, organisms)
```

# Session info

```{r}
sessionInfo()
```

