---
title: "sc-ChIPseq analysis of breast cancer"
author: "Alba Machado-Lopez"
date: "January 2023"
output: html_document
knit: (function(input_file, encoding) {
    out_dir <- 'docs';
    rmarkdown::render(input_file,
      encoding=encoding,
      output_file=file.path(dirname(input_file), out_dir, 'brca_sc_epigenomics.html'))})
---

```{r, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r echo = F}
options(error = function(){    # Beep on error
  beepr::beep()
  Sys.sleep(1)
  }
 )

.Last <- function() {          # Beep on exiting session
  beepr::beep()
  Sys.sleep(1)
  }
```


# Load libraries

```{r, message=F}
library(dplyr)
library(stringr)
library(R.utils)
library(ggplot2)
library(ggpubr)
library(Signac)
library(Seurat)
library(biomaRt)
library(harmony)
library(openxlsx)
library(GenomicRanges)
library(AnnotationHub)
theme_set(theme_bw())
```


# Obtain files

## Download data

This project uses the data from [Grosselin et al, 2019](https://pubmed.ncbi.nlm.nih.gov/31152164/9), which is uploaded to GEO.

```{bash, eval=FALSE}
mkdir -p data
cd data
wget -r -np -nd 'https://ftp.ncbi.nlm.nih.gov/geo/series/GSE117nnn/GSE117309/suppl/GSE117309_RAW.tar' -R "index.html*" # Omitting the index file
```



# Human cells
## Read files into R

First we'll read the files from scRNAseq to get relevant genes:

```{r, cache=T}
hg19_scRNA <- readRDS("data/R_objects/hg19_RNA.rds")
genes_hg19 <- rownames(hg19_scRNA)

mart.hs <- useMart("ensembl", "hsapiens_gene_ensembl")
mart_genes <- getBM(attributes = c("ensembl_gene_id", "chromosome_name", "start_position", "end_position", "hgnc_symbol"), mart = mart.hs)

hg19_coordinates <- mart_genes %>% dplyr::filter(hgnc_symbol %in% genes_hg19) %>% dplyr::filter(!grepl("CHR", chromosome_name))

# We create a genomic ranges object with the coordinates of the relevant genes
gr_genes <- GRanges(seqnames = hg19_coordinates$chromosome_name, ranges = IRanges(start = hg19_coordinates$start_position, end = hg19_coordinates$end_position, names = hg19_coordinates$hgnc_symbol))
```

Now let's read the ChIP files

```{r}
# Get a list of all files that match the patternt
ChIP_files <- list.files(path="data/",pattern=".txt", full.names = T) 

# Remove the Jurkat and Ramos files so that they are not loaded because they are not relevant to the study
scChIP_hg38_files <- ChIP_files[!grepl('Jurkat|Ramos|mm10',ChIP_files)]

# Load the files and define regions
scChIP_hg38 <- lapply(scChIP_hg38_files, read.table, header=T,row.names=1)

regions_hg38 <- data.frame(region=rownames(scChIP_hg38[[1]])) %>% mutate(chromosome_name=str_split(region,"_", simplify = T)[,1],start=str_split(region,"_", simplify = T)[,2],end=str_split(region,"_", simplify = T)[,3]) %>%
  mutate(chromosome_name=str_remove_all(chromosome_name, "chr"),start=as.numeric(start), end=as.numeric(end)) %>%
  mutate(chromosome_name=str_replace_all(chromosome_name,"M","MT")) %>% 
  dplyr::filter(!is.na(start)) %>% dplyr::filter(chromosome_name!="Y") # We remove chromosome Y as it was not present in the other dataset

gr_regions  <- GRanges(seqnames =regions_hg38$chromosome_name,
                       ranges = IRanges(start = regions_hg38$start, 
                                        end = regions_hg38$end, 
                                        names = regions_hg38$region))

# Find overlapping region
overlap_human <- findOverlaps(gr_genes, gr_regions)


regions_genes_human <- regions_hg38[subjectHits(overlap_human), ]
regions_genes_human <- unique(regions_genes_human$region)
```


Now we'll keep only relevant regions

```{r, message=F}
scChIP_hg38 <- lapply(scChIP_hg38, function(x){
  x <- x[regions_genes_human, ]
  })

names(scChIP_hg38) <- scChIP_hg38_files %>% str_remove(".txt.gz") %>% gsub("data/GSM32908.._CountTable_","",.) %>% gsub("_scChIP_H3K27me3","",.)



# Read all files into a ChromatinAssay within a Seurat Object 
import_sc_files <- function(scChip_list){
  for (i in 1:length(scChip_list)){
      chrom_assay <- CreateChromatinAssay(counts = scChip_list[[i]],
                                      sep = c("_", "_"),
                                      genome = 'hg38')
      sc_seurat <- CreateSeuratObject(counts=chrom_assay,
                                  assay="peaks",
                                  project=as.character(names(scChIP_hg38)[i]))
  scChip_list[[i]] <- sc_seurat
  }
  return(scChip_list)
}

scChIP_hg38  <- import_sc_files(scChIP_hg38)

# Merge all Seurat objects into one
scChIP_hg38_combined  <- merge(scChIP_hg38[[1]], y=c(scChIP_hg38[[2]], scChIP_hg38[[3]],scChIP_hg38[[4]]))

```

## Gene annotation 

```{r, warning=F, cache=T}
ensemble_db <- query(AnnotationHub(), c("EnsDb","Hsapiens"))[["AH104864"]]

# Get annotations from ensemble
annotations_human <- GetGRangesFromEnsDb(ensdb = ensemble_db)

# Change to UCSC style
seqlevelsStyle(annotations_human) <- 'UCSC'

# Add annotations to the seurat object
Annotation(scChIP_hg38_combined) <- annotations_human
```

## Quality control

For single-cell chromatin assays, several QC metrics have been defined, including the nucleosome binding pattern, enrichment around transcriptional start sites, total and percentage of fragments in peaks and reads in blacklist regions. Sadly, these metrics are calculated based on raw file fragments, which are not available for this project

```{r, fig.width=10, fig.height=8}
VlnPlot(scChIP_hg38_combined, 
        features = c("nFeature_peaks", "nCount_peaks"), 
        ncol = 3,
        split.by="orig.ident")
```


## QC filtering

Besides the initial filtering where we removed windows with no reads in more than 1% cells, we will now remove cells with counts below or above the 5% and 95% percentiles, respectively:

```{r}
# Define cut-offs
count_percentiles <- quantile(scChIP_hg38_combined$nCount_peaks, probs=seq(0, 1, 0.01))

# Number of reads in peaks (nCount_peaks), number of peaks that contain at least one read (nFeature_peaks)

scChIP_hg38 <- lapply(scChIP_hg38, function(x){
  x <- subset(x, subset = nCount_peaks > count_percentiles[6] & nCount_peaks < count_percentiles[96])
  return(x)
})


scChIP_hg38_combined <- subset(scChIP_hg38_combined, subset = nCount_peaks > count_percentiles[6] & nCount_peaks < count_percentiles[96])
```

## SCTransform pre-processing

```{r, message=F}
scChIP_hg38_sct <- lapply(scChIP_hg38, SCTransform, assay="peaks")
features_hg38_sct <- SelectIntegrationFeatures(object.list = scChIP_hg38_sct, nfeatures = 3792)
scChIP_hg38_sct <- PrepSCTIntegration(object.list = scChIP_hg38_sct, anchor.features = features_hg38_sct)

anchors_hg38_sct <- FindIntegrationAnchors(object.list = scChIP_hg38_sct, normalization.method = "SCT", anchor.features = features_hg38_sct)
scChIP_hg38_sct_combined <- IntegrateData(anchorset = anchors_hg38_sct, normalization.method = "SCT")

scChIP_hg38_sct_combined <- RunPCA(scChIP_hg38_sct_combined, verbose = FALSE)
scChIP_hg38_sct_combined <- RunUMAP(scChIP_hg38_sct_combined, reduction = "pca", dims = 1:30)
scChIP_hg38_sct_combined <- FindNeighbors(scChIP_hg38_sct_combined, dims = 1:30, verbose = FALSE)
scChIP_hg38_sct_combined <- FindClusters(scChIP_hg38_sct_combined, verbose = FALSE)
plot_1 <- DimPlot(scChIP_hg38_sct_combined, reduction = "umap") + ggtitle("Integrated with SCT")
```

### Without integration

```{r}
scChIP_hg38_sct <- lapply(scChIP_hg38_sct, function(x){
  x <- RunPCA(x)
  x <- FindNeighbors(x, dims = 1:30)
  x <- FindClusters(x, resolution = 0.7)
  x <- RunUMAP(x, dims = 1:30)
  return(x)
})

DimPlot(scChIP_hg38_sct[[1]], reduction = "umap")
DimPlot(scChIP_hg38_sct[[2]], reduction = "umap")
DimPlot(scChIP_hg38_sct[[3]], reduction = "umap")
DimPlot(scChIP_hg38_sct[[4]], reduction = "umap") 

```


## Signac pre-processing

### Normalization and feature selection

We will carry this process both in the list of Seurat objects and in the combined dataset for integration in the following step

```{r}
# Process the individual datasets
scChIP_hg38 <- lapply(scChIP_hg38, function(x){
  x <- RunTFIDF(x)
  x <- FindTopFeatures(x, min.cutoff = 'q0')
  x <- RunSVD(x)
  return(x)
})


# Process the combined dataset
scChIP_hg38_combined <- FindTopFeatures(scChIP_hg38_combined, min.cutoff = 10)
scChIP_hg38_combined <- RunTFIDF(scChIP_hg38_combined)
scChIP_hg38_combined <- RunSVD(scChIP_hg38_combined)
DepthCor(scChIP_hg38_combined) # Check if any of the PCs is strongly associated with region coverage

scChIP_hg38_combined <- RunUMAP(scChIP_hg38_combined, reduction = "lsi", dims = 2:30)
p1 <- DimPlot(scChIP_hg38_combined, group.by = "orig.ident")+ ggtitle("Merged with Signac")
plot_2 <- DimPlot(scChIP_hg38_combined)+ ggtitle("Merged with Signac")
```

Based on this graph we will remove the first PC from subsequent analyses

### Integration with Harmony

```{r}
scChIP_hg38_combined <- scChIP_hg38_combined %>% ScaleData(.) %>% RunHarmony(group.by.vars = "orig.ident", plot_convergence=T, dims.use=2:30, reduction="lsi", assay.use="peaks")

p2 <- DimPlot(object = scChIP_hg38_combined, reduction = "harmony", pt.size = .1, group.by = "orig.ident") + ggtitle("Integration with harmony")

plot_3 <-  DimPlot(object = scChIP_hg38_combined, reduction = "harmony", pt.size = .1) + ggtitle("Integrated with harmony")

ggarrange(p1, p2, common.legend = T, legend="right")
```


### Dimensionality reduction and clustering

```{r}
scChIP_hg38_combined <- scChIP_hg38_combined %>% 
    FindNeighbors(reduction = "harmony", dims = 2:30) %>% 
    FindClusters(resolution = 0.5, algorithm = 3) %>% 
    identity()

DimPlot(scChIP_hg38_combined, reduction = "harmony", label=T, group.by = 'orig.ident')

DimPlot(scChIP_hg38_combined, reduction = "harmony", label=T)
DimPlot(scChIP_hg38_combined, reduction = "umap", label=T)

```

### Clustering without integration

```{r}
scChIP_hg38 <- lapply(scChIP_hg38, function(x){
  x <- RunUMAP(x, reduction = "lsi", dims = 2:30) 
  x <- FindNeighbors(x, reduction = "lsi", dims = 2:30)  
  x <- FindClusters(x, resolution = 0.5, algorithm = 3)
  return(x)
})

lapply(scChIP_hg38, function(x){
  DimPlot(x, reduction = "lsi", label=T)
})

```

## Seurat pre-processing

```{r, message=F}
scChIP_hg38_seurat <- scChIP_hg38

scChIP_hg38_seurat_combined  <- merge(scChIP_hg38_seurat[[1]], y=c(scChIP_hg38_seurat[[2]], scChIP_hg38_seurat[[3]],scChIP_hg38_seurat[[4]]))

scChIP_hg38_seurat <- lapply(scChIP_hg38_seurat, function(x){
  x <- NormalizeData(x, assay="peaks")
  x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
  x <- ScaleData(x)
  x <- RunPCA(x)
  x <- FindNeighbors(x, dims = 1:30)
  x <- FindClusters(x, resolution = 0.5)
  x <- RunUMAP(x, dims = 1:30)
  return(x)
})

scChIP_hg38_seurat_combined <- NormalizeData(scChIP_hg38_seurat_combined, assay="peaks")
scChIP_hg38_seurat_combined <- ScaleData(scChIP_hg38_seurat_combined)
scChIP_hg38_seurat_combined <- RunPCA(scChIP_hg38_seurat_combined,features=rownames(scChIP_hg38_seurat_combined))
scChIP_hg38_seurat_combined <- FindNeighbors(scChIP_hg38_seurat_combined, dims = 1:50)
scChIP_hg38_seurat_combined <- FindClusters(scChIP_hg38_seurat_combined, resolution = 0.4)
scChIP_hg38_seurat_combined <- RunUMAP(scChIP_hg38_seurat_combined, dims = 1:50)


DimPlot(scChIP_hg38_seurat[[1]], reduction = "umap")
DimPlot(scChIP_hg38_seurat[[2]], reduction = "umap")
DimPlot(scChIP_hg38_seurat[[3]], reduction = "umap")
DimPlot(scChIP_hg38_seurat[[4]], reduction = "umap")

plot_4 <- DimPlot(scChIP_hg38_seurat_combined, reduction = "umap") + ggtitle("Integrated with Seurat")
DimPlot(scChIP_hg38_seurat_combined, reduction = "umap", group.by="orig.ident")
```



```{r}
features_hg38 <- SelectIntegrationFeatures(object.list = scChIP_hg38_seurat)
hg38_seurat_anchors <- FindIntegrationAnchors(object.list = scChIP_hg38_seurat, anchor.features = features_hg38)
hg38_seurat_integrated<- IntegrateData(anchorset = hg38_seurat_anchors)
```

```{r}
DefaultAssay(hg38_seurat_integrated) <- "integrated"

# Run the standard workflow for visualization and clustering
hg38_seurat_integrated <- ScaleData(hg38_seurat_integrated, verbose = FALSE)
hg38_seurat_integrated <- RunPCA(hg38_seurat_integrated, npcs = 30, verbose = FALSE)
hg38_seurat_integrated <- RunUMAP(hg38_seurat_integrated, reduction = "pca", dims = 1:30)
hg38_seurat_integrated <- FindNeighbors(hg38_seurat_integrated, reduction = "pca", dims = 1:30)
hg38_seurat_integrated <- FindClusters(hg38_seurat_integrated, resolution = 0.5)
# Visualization
p1 <- DimPlot(hg38_seurat_integrated, reduction = "umap", group.by = "orig.ident")
p2 <- DimPlot(hg38_seurat_integrated, reduction = "umap", label = TRUE, repel = TRUE)
p1 + p2
```

```{r, fig.width=10, fig.height=10}
ggarrange(plot_1, plot_2, plot_3, plot_4, common.legend = T, legend="right")
```

## Marker detection

```{r}
hg38_markers <- FindAllMarkers(hg38_seurat_integrated)
```


```{r, eval=F}
closest_genes <- list()

DefaultAssay(hg38_seurat_integrated) <- "peaks"
Annotation(hg38_seurat_integrated) <- annotations_human

for (i in c(0,1,2,3)){
  cluster_markers <- hg38_markers %>% 
    dplyr::filter(cluster==i) %>% 
    dplyr::filter(p_val_adj <0.05)
  closest_genes_new <- ClosestFeature(object=hg38_seurat_integrated, 
                                      regions = cluster_markers$gene)
  closest_genes_new$cluster <- i
  list_n <- i+1
  closest_genes[[list_n]] <- closest_genes_new}


closest_genes <- do.call(rbind, closest_genes)

```

## Integration with RNA

We will convert our data to a binary gene_activity_matrix, where 1 indicates expression of the gene (i.e no counts in the repressive histone mark) and 0 indicates no expression of the gene (i.e. counts in the repressive histone mark)

```{r}
gene_activity_matrix <- hg38_seurat_integrated@assays$integrated@data
gene_activity_matrix <- as.matrix(gene_activity_matrix)
gene_activity_matrix <- as.matrix((gene_activity_matrix<0)+0)

genes_per_window_hg38 <- cbind(hg19_coordinates[queryHits(overlap_human),], regions_hg38[subjectHits(overlap_human), ])

names(genes_per_window_hg38) <- c("ensembl_gene_id","chromosome","start_position","end_position","hgnc_symbol","region","chromosome_name","start","end")

genes_per_window_hg38 <- genes_per_window_hg38 %>% mutate(distance=abs(start-start_position),
                                                          region=str_replace_all(region,"_","-")) 

# Let's order the dataframe to remove duplicates based on the smallest distance to the genomic bin defined
genes_per_window_hg38 <- genes_per_window_hg38[order(genes_per_window_hg38[,'distance']),]
genes_per_window_hg38 <- genes_per_window_hg38[!duplicated(genes_per_window_hg38$distance),]


gene_activity_matrix <- as.data.frame(gene_activity_matrix)

gene_activity_matrix$gene <- genes_per_window_hg38$hgnc_symbol[match(rownames(gene_activity_matrix), genes_per_window_hg38$region)]

gene_activity_matrix <- aggregate(. ~ gene, median, data = gene_activity_matrix)

rownames(gene_activity_matrix) <- gene_activity_matrix$gene
gene_activity_matrix <- gene_activity_matrix %>% dplyr::select(-gene)

# add gene activities as a new assay
hg38_seurat_integrated[["ACTIVITY"]] <- CreateAssayObject(counts=gene_activity_matrix)
```


```{r}
DefaultAssay(hg38_seurat_integrated) <- "ACTIVITY"
hg38_seurat_integrated <- NormalizeData(hg38_seurat_integrated)
hg38_seurat_integrated <- ScaleData(hg38_seurat_integrated)
hg38_seurat_integrated <- RunPCA(hg38_seurat_integrated, features=rownames(hg38_seurat_integrated))


DefaultAssay(hg19_scRNA) <- "integrated"

transfer.anchors <- FindTransferAnchors(
  reference = hg19_scRNA,
  query = hg38_seurat_integrated,
  reference.reduction = "pca",k.filter = NA)

hg38_seurat_integrated <- MapQuery(
  anchorset = transfer.anchors,
  reference = hg19_scRNA,
  query = hg38_seurat_integrated,
  refdata = hg19_scRNA$cluster_name
)
```


```{r}
plot1 <- DimPlot(
  object = hg19_scRNA,
  group.by = 'cluster_name',
  label = TRUE,
  repel = TRUE) + NoLegend() + ggtitle('scRNA-seq')

plot2 <- DimPlot(
  object = hg38_seurat_integrated,
  group.by = 'predicted.id',
  label = TRUE,
  repel = TRUE)  + ggtitle('scChIP-seq')

plot1 + plot2
```

```{r}
hg38_seurat_integrated@meta.data$predicted_chip <- case_when(hg38_seurat_integrated@meta.data$predicted.id == "Fibroblasts" ~ "Fibroblasts",T~"Immune")



hg38_seurat_integrated@meta.data$states <- case_when(hg38_seurat_integrated@meta.data$seurat_clusters == 0 ~ "Chromatin state 1",
                                                     hg38_seurat_integrated@meta.data$seurat_clusters == 1 ~ "Chromatin state 2",
                                                     hg38_seurat_integrated@meta.data$seurat_clusters == 2 ~ "Chromatin state 3",
                                                     hg38_seurat_integrated@meta.data$seurat_clusters == 3 ~ "Chromatin state 4")

```


# Mouse cells
## Read files into R

First we'll read the files from scRNAseq to get relevant genes:

```{r, cache=T}
mm10_scRNA <- readRDS("data/R_objects/mm10_RNA.rds")
genes_mm10 <- rownames(mm10_scRNA)

# The next two lines are commented because biomart can be difficult to access with poor internet connection (currently my case) so I will update the RData objects stroed
# mart.mm <- useMart("ensembl", "mmusculus_gene_ensembl")
# mart_genes_mm <- getBM(attributes = c("ensembl_gene_id", "chromosome_name", "start_position", "end_position", "external_gene_name"), mart = mart.mm)

# saveRDS(mart_genes_mm,"data/R_objects/mart_genes_mm.rds")

mart_genes_mm <- readRDS("data/R_objects/mart_genes_mm.rds")

mm10_coordinates <- mart_genes_mm %>% dplyr::filter(external_gene_name %in% genes_mm10) %>% dplyr::filter(!grepl("CHR", chromosome_name))

# We create a genomic ranges object with the coordinates of the relevant genes
gr_genes <- GRanges(seqnames = mm10_coordinates$chromosome_name, ranges = IRanges(start = mm10_coordinates$start_position, end = mm10_coordinates$end_position, names = mm10_coordinates$hgnc_symbol))
```

Now let's read the ChIP files

```{r}
# Remove the Jurkat and Ramos files so that they are not loaded because they are not relevant to the study
scChIP_mm10_files <- ChIP_files[!grepl('Jurkat|Ramos|hg38',ChIP_files)]

# Load the files and define regions
scChIP_mm10 <- lapply(scChIP_mm10_files, read.table, header=T,row.names=1)

regions_mm10 <- data.frame(region=rownames(scChIP_mm10[[1]])) %>% mutate(chromosome_name=str_split(region,"_", simplify = T)[,1],start=str_split(region,"_", simplify = T)[,2],end=str_split(region,"_", simplify = T)[,3]) %>%
  mutate(chromosome_name=str_remove_all(chromosome_name, "chr"),start=as.numeric(start), end=as.numeric(end)) %>%
  mutate(chromosome_name=str_replace_all(chromosome_name,"M","MT")) %>% 
  dplyr::filter(!is.na(start)) %>% dplyr::filter(chromosome_name!="Y") # We remove chromosome Y as it was not present in the other dataset

gr_regions  <- GRanges(seqnames =regions_mm10$chromosome_name,
                       ranges = IRanges(start = regions_mm10$start, 
                                        end = regions_mm10$end, 
                                        names = regions_mm10$region))

# Find overlapping region
overlap_mm10 <- findOverlaps(gr_genes, gr_regions)


regions_genes_mm10 <- regions_mm10[subjectHits(overlap_mm10), ]
regions_genes_mm10 <- unique(regions_genes_mm10$region)
```


Now we'll keep only relevant regions

```{r, message=F}
scChIP_mm10 <- lapply(scChIP_mm10, function(x){
  x <- x[regions_genes_mm10, ]
  })

names(scChIP_mm10) <- scChIP_mm10_files %>% str_remove(".txt.gz") %>% gsub("data/GSM32908.._CountTable_","",.) %>% gsub("_scChIP_H3K27me3","",.)



# Read all files into a ChromatinAssay within a Seurat Object 
import_sc_files <- function(scChip_list){
  for (i in 1:length(scChip_list)){
      chrom_assay <- CreateChromatinAssay(counts = scChip_list[[i]],
                                      sep = c("_", "_"),
                                      genome = 'mm10')
      sc_seurat <- CreateSeuratObject(counts=chrom_assay,
                                  assay="peaks",
                                  project=as.character(names(scChIP_mm10)[i]))
  scChip_list[[i]] <- sc_seurat
  }
  return(scChip_list)
}

scChIP_mm10  <- import_sc_files(scChIP_mm10)

# Merge all Seurat objects into one
scChIP_mm10_combined  <- merge(scChIP_mm10[[1]], y=c(scChIP_mm10[[2]], scChIP_mm10[[3]],scChIP_mm10[[4]]))

```

## Gene annotation 

```{r, warning=F}
# Same issue with conecction for ensemble

#ensemble_db_mm <-  query(AnnotationHub(), c("EnsDb","Mmusculus"))[["AH89211"]] # We retrieve ensembl version 102 because it is the last one using the mm10  (GRCm38) genome


# Get annotations from ensemble
# annotations_mm10 <- GetGRangesFromEnsDb(ensdb = ensemble_db_mm)

#saveRDS(annotations_mice,"data/R_objects/annotations_mm10.rds")

annotations_mm10 <- readRDS("data/R_objects/annotations_mm10.rds")

# Change to UCSC style
seqlevelsStyle(annotations_mm10) <- 'UCSC'

# Add annotations to the seurat object
Annotation(scChIP_mm10_combined) <- annotations_mm10
```

## Quality control

For single-cell chromatin assays, several QC metrics have been defined, including the nucleosome binding pattern, enrichment around transcriptional start sites, total and percentage of fragments in peaks and reads in blacklist regions. Sadly, these metrics are calculated based on raw file fragments, which are not available for this project

```{r, fig.width=10, fig.height=8}
VlnPlot(scChIP_mm10_combined, 
        features = c("nFeature_peaks", "nCount_peaks"), 
        ncol = 3,
        split.by="orig.ident")
```


## QC filtering

Besides the initial filtering where we removed windows with no reads in more than 1% cells, we will now remove cells with counts below or above the 5% and 95% percentiles, respectively:

```{r}
# Define cut-offs
count_percentiles <- quantile(scChIP_mm10_combined$nCount_peaks, probs=seq(0, 1, 0.01))

# Number of reads in peaks (nCount_peaks), number of peaks that contain at least one read (nFeature_peaks)

scChIP_mm10 <- lapply(scChIP_mm10, function(x){
  x <- subset(x, subset = nCount_peaks > count_percentiles[6] & nCount_peaks < count_percentiles[96])
  return(x)
})


scChIP_mm10_combined <- subset(scChIP_mm10_combined, subset = nCount_peaks > count_percentiles[6] & nCount_peaks < count_percentiles[96])
```

## SCTransform pre-processing

```{r, message=F}
scChIP_mm10_sct <- lapply(scChIP_mm10, SCTransform, assay="peaks")
features_mm10_sct <- SelectIntegrationFeatures(object.list = scChIP_mm10_sct, nfeatures = 3728)
scChIP_mm10_sct <- PrepSCTIntegration(object.list = scChIP_mm10_sct, anchor.features = features_mm10_sct)

anchors_mm10_sct <- FindIntegrationAnchors(object.list = scChIP_mm10_sct, normalization.method = "SCT", anchor.features = features_mm10_sct, k.filter = NA)
scChIP_mm10_sct_combined <- IntegrateData(anchorset = anchors_mm10_sct, normalization.method = "SCT", k.weight=10)

scChIP_mm10_sct_combined <- RunPCA(scChIP_mm10_sct_combined, verbose = FALSE)
scChIP_mm10_sct_combined <- RunUMAP(scChIP_mm10_sct_combined, reduction = "pca", dims = 1:30)
scChIP_mm10_sct_combined <- FindNeighbors(scChIP_mm10_sct_combined, dims = 1:30, verbose = FALSE)
scChIP_mm10_sct_combined <- FindClusters(scChIP_mm10_sct_combined, verbose = FALSE)
DimPlot(scChIP_mm10_sct_combined, reduction = "umap", split.by = "orig.ident")
plot_5 <- DimPlot(scChIP_mm10_sct_combined, reduction = "umap") + ggtitle("Integrated with SCT")
```

### Without integration

```{r}
scChIP_mm10_sct <- lapply(scChIP_mm10_sct, function(x){
  x <- RunPCA(x)
  x <- FindNeighbors(x, dims = 1:30)
  x <- FindClusters(x, resolution = 0.7)
  x <- RunUMAP(x, dims = 1:30)
  return(x)
})

DimPlot(scChIP_mm10_sct[[1]], reduction = "umap")
DimPlot(scChIP_mm10_sct[[2]], reduction = "umap")
DimPlot(scChIP_mm10_sct[[3]], reduction = "umap")
DimPlot(scChIP_mm10_sct[[4]], reduction = "umap") 

```


## Signac pre-processing

### Normalization and feature selection

We will carry this process both in the list of Seurat objects and in the combined dataset for integration in the following step

```{r}
# Process the individual datasets
scChIP_mm10 <- lapply(scChIP_mm10, function(x){
  x <- RunTFIDF(x)
  x <- FindTopFeatures(x, min.cutoff = 'q0')
  x <- RunSVD(x)
  return(x)
})


# Process the combined dataset
scChIP_mm10_combined <- FindTopFeatures(scChIP_mm10_combined, min.cutoff = 10)
scChIP_mm10_combined <- RunTFIDF(scChIP_mm10_combined)
scChIP_mm10_combined <- RunSVD(scChIP_mm10_combined)
DepthCor(scChIP_mm10_combined) # Check if any of the PCs is strongly associated with region coverage

scChIP_mm10_combined <- RunUMAP(scChIP_mm10_combined, reduction = "lsi", dims = 2:30)
p1 <- DimPlot(scChIP_mm10_combined, group.by = "orig.ident")+ ggtitle("Merged")
plot_6 <- DimPlot(scChIP_mm10_combined)+ ggtitle("Merged with Signac")
```

Based on this graph we will remove the first PC from subsequent analyses

### Integration with Harmony

```{r}
scChIP_mm10_combined <- scChIP_mm10_combined %>% ScaleData(.) %>% RunHarmony(group.by.vars = "orig.ident", plot_convergence=T, dims.use=2:30, reduction="lsi", assay.use="peaks")

p2 <-  DimPlot(object = scChIP_mm10_combined, reduction = "harmony", pt.size = .1, group.by = "orig.ident") + ggtitle("Integrated")

plot_7 <-  DimPlot(object = scChIP_mm10_combined, reduction = "harmony", pt.size = .1) + ggtitle("Integrated with harmony")

ggarrange(p1, p2, common.legend = T, legend="right")
```


### Dimensionality reduction and clustering

```{r}
scChIP_mm10_combined <- scChIP_mm10_combined %>% 
    FindNeighbors(reduction = "harmony", dims = 2:30) %>% 
    FindClusters(resolution = 0.5, algorithm = 3) %>% 
    identity()

DimPlot(scChIP_mm10_combined, reduction = "harmony", label=T, group.by = 'orig.ident')

DimPlot(scChIP_mm10_combined, reduction = "harmony", label=T)
DimPlot(scChIP_mm10_combined, reduction = "umap", label=T)

```

### Clustering without integration

```{r}
scChIP_mm10 <- lapply(scChIP_mm10, function(x){
  x <- RunUMAP(x, reduction = "lsi", dims = 2:30) 
  x <- FindNeighbors(x, reduction = "lsi", dims = 2:30)  
  x <- FindClusters(x, resolution = 0.5, algorithm = 3)
  return(x)
})

lapply(scChIP_mm10, function(x){
  DimPlot(x, reduction = "lsi", label=T)
})

```

## Seurat pre-processing

```{r, message=F}
scChIP_mm10_seurat <- scChIP_mm10

scChIP_mm10_seurat_combined  <- merge(scChIP_mm10_seurat[[1]], y=c(scChIP_mm10_seurat[[2]], scChIP_mm10_seurat[[3]],scChIP_mm10_seurat[[4]]))

scChIP_mm10_seurat <- lapply(scChIP_mm10_seurat, function(x){
  x <- NormalizeData(x, assay="peaks")
  x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
  x <- ScaleData(x)
  x <- RunPCA(x)
  x <- FindNeighbors(x, dims = 1:30)
  x <- FindClusters(x, resolution = 0.5)
  x <- RunUMAP(x, dims = 1:30)
  return(x)
})

scChIP_mm10_seurat_combined <- NormalizeData(scChIP_mm10_seurat_combined, assay="peaks")
scChIP_mm10_seurat_combined <- ScaleData(scChIP_mm10_seurat_combined)
scChIP_mm10_seurat_combined <- RunPCA(scChIP_mm10_seurat_combined,features=rownames(scChIP_mm10_seurat_combined))
scChIP_mm10_seurat_combined <- FindNeighbors(scChIP_mm10_seurat_combined, dims = 1:50)
scChIP_mm10_seurat_combined <- FindClusters(scChIP_mm10_seurat_combined, resolution = 0.4)
scChIP_mm10_seurat_combined <- RunUMAP(scChIP_mm10_seurat_combined, dims = 1:50)


DimPlot(scChIP_mm10_seurat[[1]], reduction = "umap")
DimPlot(scChIP_mm10_seurat[[2]], reduction = "umap")
DimPlot(scChIP_mm10_seurat[[3]], reduction = "umap")
DimPlot(scChIP_mm10_seurat[[4]], reduction = "umap")

DimPlot(scChIP_mm10_seurat_combined, reduction = "umap")
DimPlot(scChIP_mm10_seurat_combined, reduction = "umap", group.by="orig.ident")

plot_8 <- DimPlot(scChIP_mm10_seurat_combined, reduction = "umap") + ggtitle("Integrated with Seurat")
```

```{r}
features_mm10 <- SelectIntegrationFeatures(object.list = scChIP_mm10_seurat)
mm10_seurat_anchors <- FindIntegrationAnchors(object.list = scChIP_mm10_seurat, anchor.features = features_mm10, k.filter=NA)
mm10_seurat_integrated<- IntegrateData(anchorset = mm10_seurat_anchors, k.weight=10)
```

```{r}
DefaultAssay(mm10_seurat_integrated) <- "integrated"

# Run the standard workflow for visualization and clustering
mm10_seurat_integrated <- ScaleData(mm10_seurat_integrated, verbose = FALSE)
mm10_seurat_integrated <- RunPCA(mm10_seurat_integrated, npcs = 30, verbose = FALSE)
mm10_seurat_integrated <- RunUMAP(mm10_seurat_integrated, reduction = "pca", dims = 1:30)
mm10_seurat_integrated <- FindNeighbors(mm10_seurat_integrated, reduction = "pca", dims = 1:30)
mm10_seurat_integrated <- FindClusters(mm10_seurat_integrated, resolution = 0.5)
# Visualization
p1 <- DimPlot(mm10_seurat_integrated, reduction = "umap", group.by = "orig.ident")
p2 <- DimPlot(mm10_seurat_integrated, reduction = "umap", label = TRUE, repel = TRUE)
p1 + p2
```

```{r, fig.width=10, fig.height=10}
ggarrange(plot_5, plot_6, plot_7, plot_8, common.legend = T, legend="right")
```

## Marker detection

```{r}
mm10_markers <- FindAllMarkers(mm10_seurat_integrated)
```


```{r, eval=F}
closest_genes <- list()

DefaultAssay(mm10_seurat_integrated) <- "peaks"
Annotation(mm10_seurat_integrated) <- annotations_mm10

for (i in c(0,1,2)){
  cluster_markers <- mm10_markers %>% 
    dplyr::filter(cluster==i) %>% 
    dplyr::filter(p_val_adj <0.05)
  closest_genes_new <- ClosestFeature(object=mm10_seurat_integrated, 
                                      regions = cluster_markers$gene)
  closest_genes_new$cluster <- i
  list_n <- i+1
  closest_genes[[list_n]] <- closest_genes_new}


closest_genes <- do.call(rbind, closest_genes)

```

## Integration with RNA

We will convert our data to a binary gene_activity_matrix, where 1 indicates expression of the gene (i.e no counts in the repressive histone mark) and 0 indicates no expression of the gene (i.e. counts in the repressive histone mark)

```{r}
gene_activity_matrix <- mm10_seurat_integrated@assays$integrated@data
gene_activity_matrix <- as.matrix(gene_activity_matrix)
gene_activity_matrix <- as.matrix((gene_activity_matrix<0)+0)

genes_per_window_mm10 <- cbind(mm10_coordinates[queryHits(overlap_mm10),], regions_mm10[subjectHits(overlap_mm10), ])

names(genes_per_window_mm10) <- c("ensembl_gene_id","chromosome","start_position","end_position","external_gene_name","region","chromosome_name","start","end")

genes_per_window_mm10 <- genes_per_window_mm10 %>% mutate(distance=abs(start-start_position),
                                                          region=str_replace_all(region,"_","-")) 

# Let's order the dataframe to remove duplicates based on the smallest distance to the genomic bin defined
genes_per_window_mm10 <- genes_per_window_mm10[order(genes_per_window_mm10[,'distance']),]
genes_per_window_mm10 <- genes_per_window_mm10[!duplicated(genes_per_window_mm10$distance),]


gene_activity_matrix <- as.data.frame(gene_activity_matrix)

gene_activity_matrix$gene <- genes_per_window_mm10$external_gene_name[match(rownames(gene_activity_matrix), genes_per_window_mm10$region)]

gene_activity_matrix <- aggregate(. ~ gene, median, data = gene_activity_matrix)

rownames(gene_activity_matrix) <- gene_activity_matrix$gene
gene_activity_matrix <- gene_activity_matrix %>% dplyr::select(-gene)

# add gene activities as a new assay
mm10_seurat_integrated[["ACTIVITY"]] <- CreateAssayObject(counts=gene_activity_matrix)
```


```{r}
DefaultAssay(mm10_seurat_integrated) <- "ACTIVITY"
mm10_seurat_integrated <- NormalizeData(mm10_seurat_integrated)
mm10_seurat_integrated <- ScaleData(mm10_seurat_integrated)
mm10_seurat_integrated <- RunPCA(mm10_seurat_integrated, features=rownames(mm10_seurat_integrated))


DefaultAssay(mm10_scRNA) <- "integrated"

transfer.anchors <- FindTransferAnchors(
  reference = mm10_scRNA,
  query = mm10_seurat_integrated,
  reference.reduction = "pca",k.filter = NA)

mm10_seurat_integrated <- MapQuery(
  anchorset = transfer.anchors,
  reference = mm10_scRNA,
  query = mm10_seurat_integrated,
  refdata = mm10_scRNA$cluster_name
)
```


```{r, fig.width=10, fig.height=7}
plot1 <- DimPlot(
  object = mm10_scRNA,
  group.by = 'cluster_name',
  label = TRUE,
  repel = TRUE) + NoLegend() + ggtitle('scRNA-seq')

plot2 <- DimPlot(
  object = mm10_seurat_integrated,
  group.by = 'predicted.id',
  label = TRUE,
  repel = TRUE)  + ggtitle('scChIP-seq')

plot1 + plot2
```

```{r}

mm10_seurat_integrated@meta.data$states <- case_when(mm10_seurat_integrated@meta.data$seurat_clusters == 0 ~ "Chromatin state 1",
                                                     mm10_seurat_integrated@meta.data$seurat_clusters == 1 ~ "Chromatin state 2",
                                                     mm10_seurat_integrated@meta.data$seurat_clusters == 2 ~ "Chromatin state 3")
```



# Save objects

```{r, eval=F}
saveRDS(hg38_seurat_integrated, "data/R_objects/hg38_ChIP.rds")
saveRDS(mm10_seurat_integrated, "data/R_objects/mm10_ChIP.rds")
```



# Session info

```{r}
sessionInfo()
```

