---
title: "sc-ChIPseq analysis of breast cancer"
author: "Alba Machado-Lopez"
date: "November 2022"
output: html_document
knit: (function(input_file, encoding) {
    out_dir <- 'docs';
    rmarkdown::render(input_file,
      encoding=encoding,
      output_file=file.path(dirname(input_file), out_dir, 'brca_sc_epigenomics.html'))})
---

```{r, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load libraries

```{r, message=F}
library(dplyr)
library(stringr)
library(R.utils)
library(ggplot2)
library(ggpubr)
library(Signac)
library(Seurat)
library(EnsDb.Hsapiens.v86)

library(harmony)
theme_set(theme_bw())
```


# Obtain files

## Download data

This project uses the data from [Grosselin et al, 2019](https://pubmed.ncbi.nlm.nih.gov/31152164/9), which is uploaded to GEO.

```{bash, eval=FALSE}
mkdir -p data
cd data
wget -r -np -nd 'https://ftp.ncbi.nlm.nih.gov/geo/series/GSE117nnn/GSE117309/suppl/GSE117309_RAW.tar' -R "index.html*" # Omitting the index file
```

# Read files into R

```{r}
# Get a list of all files that match the patternt
ChIP_files <- list.files(path="data/",pattern=".txt", full.names = T) 

# Remove the Jurkat and Ramos files so that they are not loaded because they are not relevant to the study
scChIP_hg38_files <- ChIP_files[!grepl('Jurkat|Ramos|mm10',ChIP_files)]

# Read all files into a ChromatinAssay within a Seurat Object 
import_sc_files <- function(model){
  scChip_file <- scChIP_hg38_files[grepl(pattern=model,scChIP_hg38_files)]
  counts <- read.table(scChip_file, header=T, row.names=1) 
  counts <- counts[!grepl("alt|random|chrUn", rownames(counts)),] # Removing non-canonical chromosome ensembles
  chrom_assay <- CreateChromatinAssay(counts = counts,
                                      sep = c("_", "_"),
                                      genome = 'hg38', 
                                      min.cells=ncol(counts)*0.01) # We filter here directly those regions with counts in more than 1% of the cells
  sc_seurat <- CreateSeuratObject(counts=chrom_assay,
                                  assay="peaks", project=as.character(model))
  return(sc_seurat)
}

# Define each of the models for which we have cells
models <- c("HBCx-95_sc","HBCx-95-CapaR_sc", "HBCx-22_sc","HBCx-22-TamR_sc")
scChIP_hg38  <-lapply(models, import_sc_files)

windows_hg38 <- Reduce(intersect, lapply(scChIP_hg38, rownames))

# Merge all Seurat objects into one
scChIP_hg38_combined  <- merge(scChIP_hg38[[1]], y=c(scChIP_hg38[[2]], scChIP_hg38[[3]],scChIP_hg38[[4]]))
```

```{r}
# Get a list of all files that match the patternt
ChIP_files <- list.files(path="data/",pattern=".txt", full.names = T) 

# Remove the Jurkat and Ramos files so that they are not loaded because they are not relevant to the study
scChIP_hg38_files <- ChIP_files[!grepl('Jurkat|Ramos|mm10',ChIP_files)]

# Load the files and find common windows with counts in at least 1% of cells (we do this first of all because removing features from Seurat objects can get messy)
scChIP_hg38 <- lapply(scChIP_hg38_files, read.table, header=T,row.names=1)

scChIP_hg38 <- lapply(scChIP_hg38, function(x){
  min_cells <- ncol(x)*0.01
  windows <- rowSums(x!=0) > min_cells
  x <- x[windows, ]
  x[!grepl("alt|random|chrUn",rownames(x)),]# Removing non-canonical chromosome ensembles
  })

windows_hg38 <- Reduce(intersect, lapply(scChIP_hg38, rownames))

scChIP_hg38 <- lapply(scChIP_hg38, function(x) x[windows_hg38,])
names(scChIP_hg38) <- scChIP_hg38_files %>% str_remove(".txt.gz") %>% gsub("data/GSM32908.._CountTable_","",.) %>% gsub("_scChIP_H3K27me3","",.)



# Read all files into a ChromatinAssay within a Seurat Object 
import_sc_files <- function(scChip_list){
  for (i in 1:length(scChip_list)){
      chrom_assay <- CreateChromatinAssay(counts = scChip_list[[i]],
                                      sep = c("_", "_"),
                                      genome = 'hg38')
      sc_seurat <- CreateSeuratObject(counts=chrom_assay,
                                  assay="peaks",
                                  project=as.character(names(scChIP_hg38)[i]))
  scChip_list[[i]] <- sc_seurat
  }
  return(scChip_list)
}

scChIP_hg38  <- import_sc_files(scChIP_hg38)

# Merge all Seurat objects into one
scChIP_hg38_combined  <- merge(scChIP_hg38[[1]], y=c(scChIP_hg38[[2]], scChIP_hg38[[3]],scChIP_hg38[[4]]))

```

# Gene annotation 

```{r, warning=F}
# Get annotations from ensemble
annotations_human <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86)

# Change to UCSC style
seqlevelsStyle(annotations_human) <- 'UCSC'

# Add annotations to the seurat object
Annotation(scChIP_hg38_combined) <- annotations_human
```

# Quality control

For single-cell chromatin assays, several QC metrics have been defined, including the nucleosome binding pattern, enrichment around transcriptional start sites, total and percentage of fragments in peaks and reads in blacklist regions. Sadly, these metrics are calculated based on raw file fragments, which are not available for this project

```{r, fig.width=10, fig.height=8}
VlnPlot(scChIP_hg38_combined, 
        features = c("nFeature_peaks", "nCount_peaks"), 
        ncol = 3,
        split.by="orig.ident")
```



# QC filtering

Besides the initial filtering where we removed windows with no reads in more than 1% cells, we will now remove cells with counts below or above the 5% and 95% percentiles, respectively:

```{r}
# Define cut-offs
count_percentiles <- quantile(scChIP_hg38_combined$nCount_peaks, probs=seq(0, 1, 0.01))

# Number of reads in peaks (nCount_peaks), number of peaks that contain at least one read (nFeature_peaks)

scChIP_hg38 <- lapply(scChIP_hg38, function(x){
  x <- subset(x, subset = nCount_peaks > count_percentiles[6] & nCount_peaks < count_percentiles[96])
  return(x)
})


scChIP_hg38_combined <- subset(scChIP_hg38_combined, subset = nCount_peaks > count_percentiles[6] & nCount_peaks < count_percentiles[96])
```

# Normalization and feature selection

We will carry this process both in the list of Seurat objects and in the combined dataset for integration in the following step

```{r}
# Process the individual datasets
scChIP_hg38 <- lapply(scChIP_hg38, function(x){
  x <- RunTFIDF(x)
  x <- FindTopFeatures(x, min.cutoff = 'q0')
  x <- RunSVD(x)
  return(x)
})


# Process the combined dataset
scChIP_hg38_combined <- FindTopFeatures(scChIP_hg38_combined, min.cutoff = 10)
scChIP_hg38_combined <- RunTFIDF(scChIP_hg38_combined)
scChIP_hg38_combined <- RunSVD(scChIP_hg38_combined)
DepthCor(scChIP_hg38_combined) # Check if any of the PCs is strongly associated with region coverage

scChIP_hg38_combined <- RunUMAP(scChIP_hg38_combined, reduction = "lsi", dims = 2:30)
p1 <- DimPlot(scChIP_hg38_combined, group.by = "orig.ident")
```

Based on this graph we will remove the first PC from subsequent analyses

# Integration

```{r}
scChIP_hg38_combined <- ScaleData(scChIP_hg38_combined, verbose = F)
scChIP_hg38_combined <- RunPCA(scChIP_hg38_combined, npcs = 30)
scChIP_hg38_combined <- scChIP_hg38_combined %>% RunHarmony("orig.ident", plot_convergence=T)

p2 <-  DimPlot(object = scChIP_hg38_combined, reduction = "harmony", pt.size = .1, group.by = "orig.ident")

(p1 + ggtitle("Merged")) | (p2 + ggtitle("Integrated"))
```


# Dimensionality reduction and clustering

```{r}
scChIP_hg38_combined <- scChIP_hg38_combined %>% 
    RunUMAP(reduction = "harmony", dims = 1:30) %>% 
    FindNeighbors(reduction = "harmony", dims = 1:30) %>% 
    FindClusters(resolution = 0.5) %>% 
    identity()

DimPlot(scChIP_hg38_combined, reduction = "umap", label=T)

```


# Marker detection

```{r}
hg38_markers <- FindAllMarkers(
  object = scChIP_hg38_combined
  #test.use = 'LR',
  #latent.vars = 'peak_region_fragments'
)

open_cd4naive <- rownames(da_peaks[da_peaks$avg_log2FC > 3, ])
open_cd14mono <- rownames(da_peaks[da_peaks$avg_log2FC < -3, ])

closest_genes_cd4naive <- ClosestFeature(pbmc, regions = open_cd4naive)
closest_genes_cd14mono <- ClosestFeature(pbmc, regions = open_cd14mono)
```


# Find genes
```{r}
# set plotting order
levels(pbmc) <- c("CD4 Naive","CD4 Memory","CD8 Naive","CD8 Effector","DN T","NK CD56bright","NK CD56Dim","pre-B",'pro-B',"pDC","DC","CD14 Mono",'CD16 Mono')

CoveragePlot(
  object = scChIP_hg38_combined,
  region = rownames(scChIP_hg38_combined)[1],
  extend.upstream = 40000,
  extend.downstream = 20000
)
```



# Session info

```{r}
sessionInfo()
```

