---
title: "sc-RNAseq analysis of breast cancer"
author: "Alba Machado-Lopez"
date: "October 2022"
output: html_document
knit: (function(input_file, encoding) {
    out_dir <- 'docs';
    rmarkdown::render(input_file,
      encoding=encoding,
      output_file=file.path(dirname(input_file), out_dir, 'brca_transcriptomics.html'))})
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load libraries

```{r, message=F}
library(R.utils)
library(dplyr)
library(stringr)
library(Seurat)
library(purrr)
library(scCATCH)
library(ggplot2)
library(SingleR)
library(msigdbr)
library(tibble)
library(fgsea)
library(clusterProfiler)
library(org.Hs.eg.db)
library(org.Mm.eg.db)
library(ggpubr)
theme_set(theme_bw())
```


# Obtain files

## Download data

This project uses the data from [Grosselin et al, 2019](https://pubmed.ncbi.nlm.nih.gov/31152164/9), which is uploaded to GEO.

```{bash, eval=FALSE}
mkdir -p data
cd data
wget -r -np -nd 'https://ftp.ncbi.nlm.nih.gov/geo/series/GSE117nnn/GSE117309/suppl/GSE117309_RAW.tar' -R "index.html*" # Omitting the index file
```


## Decompress files

```{r, eval=FALSE}
# Untar the main file
untar("data/GSE117309_RAW.tar", exdir="data/")

# Untar sc-RNAseq files
RNA_files <- list.files(path="data/",pattern="scRNA")
untar_RNA_files <- function(file){untar(paste0("data/",file), exdir="data/RNA_files")}
sapply(RNA_files, untar_RNA_files)
```

# Read files into R

```{r}
# Create function to load the files
load_scRNA_files <- function(model, organism){
  directory <- paste0("data/RNA_files/filtered_gene_bc_matrices_",model,"/",organism,"/")
  expression_matrix <- ReadMtx(
  mtx=paste0(directory, "matrix.mtx"),
  features=paste0(directory, "genes.tsv"),
  cells=paste0(directory, "barcodes.tsv"))
    # Correct gene names for easier automatization
  expression_matrix@Dimnames[[1]] <- str_remove_all(expression_matrix@Dimnames[[1]],
                                             pattern=paste0(organism,"_"))
  seurat_object <- CreateSeuratObject(counts = expression_matrix, project=model)

  return(seurat_object)
}

models <- c("HBCx-95","HBCx-95_CAPAR", "HBCx-22","HBCx22-TAMR")

scRNA_files_hg19 <- mapply(FUN=load_scRNA_files, models, organism="hg19")
names(scRNA_files_hg19) <- paste0(names(scRNA_files_hg19), "_hg19")

scRNA_files_mm10 <- mapply(FUN=load_scRNA_files, models, organism="mm10")
names(scRNA_files_mm10) <- paste0(names(scRNA_files_mm10), "_mm10")
```


# Analysis of human cells 


## Data integration

```{r}
# Normalize and identify variable features
scRNA_files_hg19 <- lapply(X = scRNA_files_hg19, FUN = function(x) {
    x <- NormalizeData(x)
    x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
})

# Select features repeatedly variable between the 4 datasets
features <- SelectIntegrationFeatures(object.list = scRNA_files_hg19)

# Select anchors for integration
brca_anchors <- FindIntegrationAnchors(object.list = scRNA_files_hg19, anchor.features = features)

# Create the integrated data assay
scRNA_combined <- IntegrateData(anchorset=brca_anchors)

```

## Quality control

```{r}
# Use the integrated assay as default
DefaultAssay(scRNA_combined) <- "RNA"

# Creating a new slot with the percentage of counts on mitochondrial reads


scRNA_combined[["percent.mt"]] <- PercentageFeatureSet(scRNA_combined, pattern="MT-")

# Violin plots of specific features
VlnPlot(scRNA_combined, 
        features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), 
        ncol = 3,
        split.by="orig.ident")


```

We will remove cells with a percentage of mitochondrial reads too high (>5%) and with an abnormally high or low number of feature counts, that can suggest empty droplets or multiplets in the microfluidic system, respectively.

```{r}
scRNA_combined <- subset(scRNA_combined, subset = nFeature_RNA > 200 & nFeature_RNA < 6000 & percent.mt < 10)

VlnPlot(scRNA_combined, 
        features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), 
        ncol = 3,
        split.by="orig.ident")
```



## Visualization and clustering

### Check integration

First we will explore the UMAPs without integration
 
```{r}
# First let's do the same analysis without integration
scRNA_combined <- FindVariableFeatures(scRNA_combined)
scRNA_combined <- ScaleData(scRNA_combined, verbose = FALSE)
scRNA_combined <- RunPCA(scRNA_combined, npcs = 30, verbose = FALSE)
scRNA_combined <- RunUMAP(scRNA_combined, reduction = "pca", dims = 1:30)
scRNA_combined <- FindNeighbors(scRNA_combined, reduction = "pca", dims = 1:30)
scRNA_combined <- FindClusters(scRNA_combined, resolution = 0.5)

p1_hg38 <- DimPlot(scRNA_combined, reduction = "umap", group.by = "orig.ident") + ggtitle("hg19 pre-integration")
p2 <- DimPlot(scRNA_combined, reduction = "umap", label = TRUE, repel = TRUE)
p1_hg38 + p2


# Change the default to the integrated dataset
DefaultAssay(scRNA_combined) <- "integrated"

scRNA_combined <- ScaleData(scRNA_combined, verbose = FALSE)
scRNA_combined <- RunPCA(scRNA_combined, npcs = 30, verbose = FALSE)
scRNA_combined <- RunUMAP(scRNA_combined, reduction = "pca", dims = 1:30)
scRNA_combined <- FindNeighbors(scRNA_combined, reduction = "pca", dims = 1:30)
scRNA_combined <- FindClusters(scRNA_combined, resolution = 0.5)


# Visualization
p2_hg38 <- DimPlot(scRNA_combined, reduction = "umap", group.by = "orig.ident") + ggtitle("hg19 post-integration")
p3 <- DimPlot(scRNA_combined, reduction = "umap", label = TRUE, repel = TRUE)

```

After normalization and integration, the UMAPs of all subsets look quite similar, so we can assume integration is correct and we can proceed forward. Once this has been verified, we can go back a few steps and refine our analysis

### Determine the  dimensionality of the dataset

```{r}
scRNA_combined <- JackStraw(scRNA_combined, num.replicate = 100)
scRNA_combined <- ScoreJackStraw(scRNA_combined, dims=1:20)

JackStrawPlot(scRNA_combined, dims=1:20)
ElbowPlot(scRNA_combined)
```

Based on these two graphs, we may consider using 13-15 PCs.

```{r, fig.width=10, fig,height=8}
scRNA_combined <- FindNeighbors(scRNA_combined, reduction = "pca", dims = 1:30)
scRNA_combined <- FindClusters(scRNA_combined, resolution = 0.5) # Try different levels of resolution and chose the one that results in cluster with most biological sense
scRNA_combined <- RunUMAP(scRNA_combined, dims = 1:30)
DimPlot(scRNA_combined, reduction="umap")
DimPlot(scRNA_combined, reduction = "umap", split.by = "orig.ident") + ggtitle("Clustering in hg19 cells") + theme(plot.title = element_text(hjust=0.5))
```

## Finding marker genes

```{r}
hg19_markers <- FindAllMarkers(scRNA_combined, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
hg19_markers %>%
    group_by(cluster) %>%
    slice_max(n = 2, order_by = avg_log2FC)
```

```{r, fig.width=10, fig.height=15}
hg19_markers %>%
    group_by(cluster) %>%
    top_n(n = 5, wt = avg_log2FC) -> top5
DoHeatmap(scRNA_combined, features = top5$gene) + NoLegend()
```


## Functional analysis

### Gene set enrichment analysis

```{r, eval=F}
MSigDB_df <- msigdbr(species = "Homo sapiens")
fgsea_sets<- MSigDB_df %>% split(x = .$gene_symbol, f = .$gs_name)



plot_GSEA <- function(cluster_number){
  genes <- hg19_markers %>% filter(cluster==as.character(cluster_number)) %>%
  arrange(desc(p_val_adj)) %>% 
  dplyr::select(gene, avg_log2FC)
  
  # Create dataframe with genes
  ranks <- deframe(genes)
  
  # Perform GSEA analysis
  fgseaRes <- fgsea(fgsea_sets,
                    stats = ranks,
                    minSize=10,
                    maxSize=500,
                    nperm=1000000)
  
  # Plot
  ggplot(fgseaRes %>% filter(padj < 0.01) %>% head(n= 20), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill= NES < 7.5)) +
  coord_flip() +
  labs(x="Pathway",
       y="Normalized Enrichment Score",
       title=paste0("Hallmark pathways NES from GSEA for cluster",as.character(cluster_number)))
}

for (cluster in 0:5){plot_GSEA(cluster)}

```

### Gene ontology

```{r}
plot_GO <- function(cluster_number){
  genes <- hg19_markers %>% filter(cluster==as.character(cluster_number))
  
  # Calculate enrichment
  enrichment <- enrichGO(gene = genes$gene,
             OrgDb  = org.Hs.eg.db,
             keyType = "SYMBOL",
             ont  = "BP",
             pAdjustMethod = "BH",
             pvalueCutoff  = 0.01,
             qvalueCutoff  = 0.05)
  enrichment <- clusterProfiler::simplify(enrichment, cutoff=0.7, by="p.adjust", select_fun=min)
  
  GO_ggdata <- enrichment %>%
   as_data_frame() %>%
   arrange(Count)
  GO_ggdata$Description <- factor(GO_ggdata$Description, levels = GO_ggdata$Description)
  print(tail(GO_ggdata))

ggplot(GO_ggdata, aes(x = Description, y = Count, fill = p.adjust)) +
 geom_bar(stat = "identity") +
 scale_colour_viridis_d(begin=0,end=1) +
 coord_flip() +
 ylab("Number of genes") +
 xlab("GO Terms") +
 theme(axis.text.y = element_text(size=10))
}

for (cluster in 0:5){print(plot_GO(cluster))}
```




## Automated cluster annotation

### Method 1- scCATCH

```{r}

hg19_geneinfo <- rev_gene(data = scRNA_combined[['integrated']]@data,
                          data_type = "data",
                          species="Human",
                          geneinfo = geneinfo)

hg19_ann_1 <- createscCATCH(data = scRNA_combined[['integrated']]@data, 
                          cluster = as.character(Idents(scRNA_combined)))

hg19_ann_1 <- findmarkergene(object = hg19_ann_1, species = "Human", marker = cellmatch, tissue = "Breast", cancer="Breast Cancer", use_method = "2")

hg19_ann_1 <- findcelltype(hg19_ann_1)
hg19_ann_1@celltype 
```
### Method 2- SingleR

```{r}
hpca.se <- HumanPrimaryCellAtlasData()
hpca.se

hg19_ann_2 <- SingleR(test = scRNA_combined[['integrated']]@data, 
                      ref = hpca.se, 
                      assay.type.test=1,
                      labels = hpca.se$label.main, clusters = scRNA_combined@meta.data[["seurat_clusters"]])

hg19_ann_2$labels
```

## Cell population labeling

```{r, fig.width=10, fig,height=8}
levels(scRNA_combined)
new_clusters_hg19 <- c("Fibroblasts","T-cells","Helper-T-Cells 1","NK cells","Helper-T-Cells 2")

names(new_clusters_hg19) <- levels(scRNA_combined)

scRNA_combined <- RenameIdents(scRNA_combined, new_clusters_hg19)

DimPlot(scRNA_combined, reduction="umap", label=T, label.box =T) 
```

# Analysis of mouse cells 

## Data integration

```{r}
# Normalize and identify variable features
scRNA_files_mm10 <- lapply(X = scRNA_files_mm10, FUN = function(x) {
    x <- NormalizeData(x)
    x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
})

# Select features repeatedly variable between the 4 datasets
features_mm10 <- SelectIntegrationFeatures(object.list = scRNA_files_mm10)

# Select anchors for integration
brca_anchors_mm10 <- FindIntegrationAnchors(object.list = scRNA_files_mm10, anchor.features = features_mm10)

# Create the integrated data assay
scRNA_combined_mm10 <- IntegrateData(anchorset=brca_anchors_mm10)

```

## Quality control

```{r}
# Use the integrated assay as default
DefaultAssay(scRNA_combined_mm10) <- "RNA"

# Creating a new slot with the percentage of counts on mitochondrial reads


scRNA_combined_mm10[["percent.mt"]] <- PercentageFeatureSet(scRNA_combined_mm10, pattern="mt-")

# Violin plots of specific features
VlnPlot(scRNA_combined_mm10, 
        features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), 
        ncol = 3,
        split.by="orig.ident")


```

We will remove cells with a percentage of mitochondrial reads too high (>5%) and with an abnormally high or low number of feature counts, that can suggest empty droplets or multiplets in the microfluidic system, respectively.

```{r}
scRNA_combined_mm10 <- subset(scRNA_combined_mm10, subset = nFeature_RNA > 200 & nFeature_RNA < 6000 & percent.mt < 10)

VlnPlot(scRNA_combined_mm10, 
        features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), 
        ncol = 3,
        split.by="orig.ident")
```



## Visualization and clustering

### Check integration

```{r}
# First let's do the same analysis without integration
scRNA_combined_mm10 <- FindVariableFeatures(scRNA_combined_mm10)
scRNA_combined_mm10 <- ScaleData(scRNA_combined_mm10, verbose = FALSE)
scRNA_combined_mm10 <- RunPCA(scRNA_combined_mm10, npcs = 30, verbose = FALSE)
scRNA_combined_mm10 <- RunUMAP(scRNA_combined_mm10, reduction = "pca", dims = 1:30)
scRNA_combined_mm10 <- FindNeighbors(scRNA_combined_mm10, reduction = "pca", dims = 1:30)
scRNA_combined_mm10 <- FindClusters(scRNA_combined_mm10, resolution = 0.5)

p1_mm10 <- DimPlot(scRNA_combined_mm10, reduction = "umap", group.by = "orig.ident") + ggtitle("mm10 pre-integration")
p2 <- DimPlot(scRNA_combined_mm10, reduction = "umap", label = TRUE, repel = TRUE)
p1_mm10 + p2

DimPlot(scRNA_combined_mm10, reduction = "umap", split.by = "orig.ident")

# Change the default to the integrated dataset
DefaultAssay(scRNA_combined_mm10) <- "integrated"

scRNA_combined_mm10 <- ScaleData(scRNA_combined_mm10, verbose = FALSE)
scRNA_combined_mm10 <- RunPCA(scRNA_combined_mm10, npcs = 30, verbose = FALSE)
scRNA_combined_mm10 <- RunUMAP(scRNA_combined_mm10, reduction = "pca", dims = 1:30)
scRNA_combined_mm10 <- FindNeighbors(scRNA_combined_mm10, reduction = "pca", dims = 1:30)
scRNA_combined_mm10 <- FindClusters(scRNA_combined_mm10, resolution = 0.5)


# Visualization
p2_mm10 <- DimPlot(scRNA_combined_mm10, reduction = "umap", group.by = "orig.ident") + ggtitle("mm10 post-integration")
p3 <- DimPlot(scRNA_combined_mm10, reduction = "umap", label = TRUE, repel = TRUE)
p2_mm10 + p3


DimPlot(scRNA_combined_mm10, reduction = "umap", split.by = "orig.ident")
```

After normalization and integration, the UMAPs of all subsets look quite similar, so we can assume integration is correct and we can proceed forward. Once this has been verified, we can go back a few steps and refine our analysis

### Determine the  dimensionality of the dataset

```{r}
scRNA_combined_mm10 <- JackStraw(scRNA_combined_mm10, num.replicate = 100)
scRNA_combined_mm10 <- ScoreJackStraw(scRNA_combined_mm10, dims=1:20)

JackStrawPlot(scRNA_combined_mm10, dims=1:20)
ElbowPlot(scRNA_combined_mm10)
```

Based on these two graphs, we may consider using 10 PCs.

```{r, fig.width=10, fig,height=8}
scRNA_combined_mm10 <- FindNeighbors(scRNA_combined_mm10, reduction = "pca", dims = 1:30)
scRNA_combined_mm10 <- FindClusters(scRNA_combined_mm10, resolution = 0.5) # Try different levels of resolution and chose the one that results in cluster with most biological sense1
scRNA_combined_mm10 <- RunUMAP(scRNA_combined_mm10, dims = 1:10)
DimPlot(scRNA_combined_mm10, reduction="umap")
DimPlot(scRNA_combined_mm10, reduction = "umap", split.by = "orig.ident") + ggtitle("Clustering in mm10 cells") + theme(plot.title = element_text(hjust=0.5))
```

## Finding marker genes

```{r}
mm10_markers <- FindAllMarkers(scRNA_combined_mm10, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
mm10_markers %>%
    group_by(cluster) %>%
    slice_max(n = 2, order_by = avg_log2FC)
```

```{r, fig.width=13, fig.height=15}
mm10_markers %>%
    group_by(cluster) %>%
    top_n(n = 5, wt = avg_log2FC) -> top5
DoHeatmap(scRNA_combined_mm10, features = top5$gene) + NoLegend()
```


## Functional analysis

### Gene set enrichment analysis

```{r, eval=F}
MSigDB_df <- msigdbr(species = "Homo sapiens")
fgsea_sets<- MSigDB_df %>% split(x = .$gene_symbol, f = .$gs_name)



plot_GSEA <- function(cluster_number){
  genes <- mm10_markers %>% filter(cluster==as.character(cluster_number)) %>%
  arrange(desc(p_val_adj)) %>% 
  dplyr::select(gene, avg_log2FC)
  
  # Create dataframe with genes
  ranks <- deframe(genes)
  
  # Perform GSEA analysis
  fgseaRes <- fgsea(fgsea_sets,
                    stats = ranks,
                    minSize=10,
                    maxSize=500,
                    nperm=1000000)
  
  # Plot
  ggplot(fgseaRes %>% filter(padj < 0.01) %>% head(n= 20), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill= NES < 7.5)) +
  coord_flip() +
  labs(x="Pathway",
       y="Normalized Enrichment Score",
       title=paste0("Hallmark pathways NES from GSEA for cluster",as.character(cluster_number)))
}

plot_GSEA(0)

for (cluster in 0:5){plot_GSEA(cluster)}

```

### Gene ontology

```{r}
plot_GO_mm10 <- function(cluster_number){
  genes <- mm10_markers %>% filter(cluster==as.character(cluster_number))
  
  # Calculate enrichment
  enrichment <- enrichGO(gene = genes$gene,
             OrgDb  = org.Mm.eg.db,
             keyType = "SYMBOL",
             ont  = "BP",
             pAdjustMethod = "BH",
             pvalueCutoff  = 0.01,
             qvalueCutoff  = 0.05)
  enrichment <- clusterProfiler::simplify(enrichment, cutoff=0.7, by="p.adjust", select_fun=min)
  
  GO_ggdata <- enrichment %>%
   as_data_frame() %>%
   arrange(Count)
  GO_ggdata$Description <- factor(GO_ggdata$Description, levels = GO_ggdata$Description)
  print(tail(GO_ggdata))
  
ggplot(GO_ggdata, aes(x = Description, y = Count, fill = p.adjust)) +
 geom_bar(stat = "identity") +
 scale_colour_viridis_d(begin=0,end=1) +
 coord_flip() +
 ylab("Number of genes") +
 xlab("GO Terms") +
 theme(axis.text.y = element_text(size=10))
}

for (cluster in 0:13){print(plot_GO_mm10(cluster))}
```


## Automated cluster annotation

### Method 1- scCATCH

```{r}
MouseRNAseqData(ensembl = FALSE, cell.ont = c("all", "nonna", "none"))

mm10_geneinfo <- rev_gene(data = scRNA_combined_mm10[['integrated']]@data,
                          data_type = "data",
                          species="Mouse",
                          geneinfo = geneinfo)

mm10_ann_1 <- createscCATCH(data = scRNA_combined_mm10[['integrated']]@data, 
                          cluster = as.character(Idents(scRNA_combined_mm10)))

mm10_ann_1 <- findmarkergene(object = mm10_ann_1, species = "Mouse", marker = cellmatch, tissue="Mammary gland", use_method = "2")

mm10_ann_1 <- findcelltype(mm10_ann_1)
mm10_ann_1@celltype
```

### Method 2- SingleR

```{r}
mRNAseq.se <- MouseRNAseqData()
mRNAseq.se

mm10_ann_2 <- SingleR(test = scRNA_combined_mm10[['integrated']]@data, 
                      ref = mRNAseq.se, 
                      assay.type.test=1,
                      labels = mRNAseq.se$label.main,
                      clusters = scRNA_combined_mm10@meta.data[["seurat_clusters"]])

mm10_ann_2
```

## Cell population labeling

```{r}
new_clusters_mm10 <- c("Fibroblasts 1","Macrophages 1","Fibroblasts 2","Fibroblasts 3","Fibroblasts 4","Endothelial cells","Fibroblasts 5","Monocytes 1","Fibroblasts 6"," NK cells","Fibroblasts 7","Macrophages 2","Macrophages 3","Monocytes 2")

names(new_clusters_mm10) <- levels(scRNA_combined_mm10)

scRNA_combined_mm10 <- RenameIdents(scRNA_combined_mm10, new_clusters_mm10)

DimPlot(scRNA_combined_mm10, reduction="umap", label=T, label.box=F, label.size = 3.5) 
```

# Figure preparation 

## Integration before and after for hg19 and mm10
```{r, fig.width=10, fig.height=10}

ggarrange(p1_hg38, p2_hg38, p1_mm10, p2_mm10, common.legend = T, legend="right", labels=c("A","B","C","D"))
```

# Save data 

We will store the relevant objects to use them in other analyses

```{r}
saveRDS(scRNA_combined, "data/R_objects/hg19_RNA.rds")
saveRDS(scRNA_combined_mm10, "data/R_objects/mm10_RNA.rds")
```


# Session info

```{r}
sessionInfo()
```

