---
title: "Comparative transcriptomic analysis of breast cancer at single cell level"
author: "Alba Machado-Lopez"
date: "December 2022"
output: html_document
knit: (function(input_file, encoding) {
    out_dir <- 'docs';
    rmarkdown::render(input_file,
      encoding=encoding,
      output_file=file.path(dirname(input_file), out_dir, 'comparative_transcriptomics.html'))})
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load libraries

```{r, message=F}
library(miloR)
library(Seurat)
library(SingleCellExperiment)
library(scater)
library(scuttle)
library(scran)
library(org.Hs.eg.db)
library(org.Mm.eg.db)
library(clusterProfiler)
library(msigdbr)
library(dplyr)
library(ggplot2)
library(patchwork)
library(ggpubr)
theme_set(theme_bw())
```


# Load objects

We will simply read the files we created with the combined seurat objects from human and mouse cells. 

```{r}
hg19_scRNA <- readRDS("data/R_objects/hg19_RNA.rds")
mm10_scRNA <- readRDS("data/R_objects/mm10_RNA.rds")
```

# Add metadata

We will now add metadata columns indicating the subtype of breast cancer and whether the tumor is resistant or sensitive to chemotherapy

```{r}
hg19_scRNA@meta.data <- hg19_scRNA@meta.data %>%
  mutate(subtype=case_when(
    orig.ident=="HBCx-95"| orig.ident=="HBCx-95_CAPAR"~"Triple_negative",
    orig.ident=="HBCx-22"| orig.ident=="HBCx22-TAMR"~"Luminal")) %>% 
  mutate(drug_resistance=case_when(
    grepl("R", orig.ident)==T~"Resistant",
    grepl("R", orig.ident)==F~"Sensitive"),
    sample=rownames(.))

mm10_scRNA@meta.data <- mm10_scRNA@meta.data %>%
  mutate(subtype=case_when(
    orig.ident=="HBCx-95"| orig.ident=="HBCx-95_CAPAR"~"Triple_negative",
    orig.ident=="HBCx-22"| orig.ident=="HBCx22-TAMR"~"Luminal")) %>% 
  mutate(drug_resistance=case_when(
    grepl("R", orig.ident)==T~"Resistant",
    grepl("R", orig.ident)==F~"Sensitive"),
    sample=rownames(.))
```


# Create Milo object

```{r}

# Create object that can store the KNN graph as a reduced dimension
hg19_scRNA_milo <- as.SingleCellExperiment(hg19_scRNA)
hg19_scRNA_milo <- Milo(hg19_scRNA_milo)

# Same for mouse
mm10_scRNA_milo <- as.SingleCellExperiment(mm10_scRNA)
mm10_scRNA_milo <- Milo(mm10_scRNA_milo)
```


# Construct KNN graph

```{r}
hg19_scRNA_milo <- buildGraph(hg19_scRNA_milo, k=20, d=30)
mm10_scRNA_milo <- buildGraph(mm10_scRNA_milo, k=30, d=30)
```

## Define neighbourhoods

```{r}
hg19_scRNA_milo <- makeNhoods(hg19_scRNA_milo, prop = 0.2, k = 20, d=30, refined = TRUE)
plotNhoodSizeHist(hg19_scRNA_milo)

mm10_scRNA_milo <- makeNhoods(mm10_scRNA_milo, prop = 0.2, k = 30, d=30, refined = TRUE)
plotNhoodSizeHist(mm10_scRNA_milo)
```

Note that we have selected k for constructing the graph and defining neighbours to achieve a distribution peak between 50 and 100 

## Count the number of cells in neighbourhoods

```{r}
hg19_scRNA_milo <- countCells(hg19_scRNA_milo, meta.data = data.frame(colData(hg19_scRNA_milo)), sample="orig.ident")

mm10_scRNA_milo <- countCells(mm10_scRNA_milo, meta.data = data.frame(colData(mm10_scRNA_milo)), sample="orig.ident")
```

# Differential abundance testing

In this step we will perform a statistical test to find populations that are significantly more or less abundant in one group in comparison to another. 

- For human cells:

```{r}
# Create design matrix
design_hg19 <- data.frame(colData(hg19_scRNA_milo))[,c("subtype", "drug_resistance","orig.ident")]
design_hg19 <- distinct(design_hg19)
rownames(design_hg19) <- design_hg19[,"orig.ident"]
design_hg19

# Calculate distances
hg19_scRNA_milo <- calcNhoodDistance(hg19_scRNA_milo, d=30)

# Perform the statistical test
comparison_subtype <- testNhoods(hg19_scRNA_milo, design = ~ drug_resistance+subtype, design.df = design_hg19)
comparison_resistance <- testNhoods(hg19_scRNA_milo, design = ~subtype + drug_resistance, design.df = design_hg19)


```

- Same approach for murine cells:

```{r}
# Create design matrix
design_mm10 <- data.frame(colData(mm10_scRNA_milo))[,c("subtype", "drug_resistance","orig.ident")]
design_mm10 <- distinct(design_mm10)
rownames(design_mm10) <- design_mm10[,"orig.ident"]
design_mm10

# Calculate distances
mm10_scRNA_milo <- calcNhoodDistance(mm10_scRNA_milo, d=30)

# Perform the statistical test
comparison_subtype_mm10 <- testNhoods(mm10_scRNA_milo, design = ~ drug_resistance+subtype, design.df = design_mm10)
comparison_resistance_mm10 <- testNhoods(mm10_scRNA_milo, design = ~ subtype+drug_resistance, design.df = design_mm10)

```

## Check that the test was balanced

```{r, fig.width=10, fig.height=6}
F1_A <- ggplot(comparison_subtype, aes(PValue)) + geom_histogram(bins=50) + ggtitle("P-value distribution for contrasts by subtype")
F1_B <- ggplot(comparison_resistance, aes(PValue)) + geom_histogram(bins=50) + ggtitle("P-value distribution for contrasts by drug sensitivity")

ggarrange(F1_A, F1_B)



F4_A <-ggplot(comparison_subtype_mm10, aes(PValue)) + geom_histogram(bins=50) + ggtitle("P-value distribution for contrasts by subtype")
F4_B <-ggplot(comparison_resistance_mm10, aes(PValue)) + geom_histogram(bins=50) + ggtitle("P-value distribution for contrasts by drug sensitivity")
ggarrange(F4_A, F4_B)
```

## Volcano plots of neighbourhoods

```{r}
ggplot(comparison_subtype, aes(logFC, -log10(SpatialFDR))) +
  geom_point() +
  geom_hline(yintercept = 1)

ggplot(comparison_resistance, aes(logFC, -log10(SpatialFDR))) +
  geom_point() +
  geom_hline(yintercept = 1)

ggplot(comparison_subtype_mm10, aes(logFC, -log10(SpatialFDR))) +
  geom_point() +
  geom_hline(yintercept = 1)

ggplot(comparison_resistance_mm10, aes(logFC, -log10(SpatialFDR))) +
  geom_point() +
  geom_hline(yintercept = 1)
```

```{r} 
hg19_scRNA_milo <- buildNhoodGraph(hg19_scRNA_milo)

## Plot single-cell UMAP
umap_hg19 <- plotReducedDim(hg19_scRNA_milo, dimred = "UMAP", colour_by="cluster_name", text_by = "cluster_name", text_size = 3) + NoLegend()
## Plot neighbourhood graph
nh_graph_hg19 <- plotNhoodGraphDA(hg19_scRNA_milo, comparison_subtype, layout="UMAP",alpha=0.05)
nh_graph_resistance_hg19 <- plotNhoodGraphDA(hg19_scRNA_milo, comparison_resistance, layout="UMAP",alpha=0.05)
```

```{r}
comparison_subtype <- annotateNhoods(hg19_scRNA_milo, comparison_subtype, coldata_col = "cluster_name")
head(comparison_subtype)
beeswarm_hg19 <- plotDAbeeswarm(comparison_subtype, group.by = "cluster_name") + theme(text = element_text(size = 10)) 

comparison_resistance <- annotateNhoods(hg19_scRNA_milo, comparison_resistance, coldata_col = "cluster_name")
head(comparison_resistance)
beeswarm_drug_r_hg19 <- plotDAbeeswarm(comparison_resistance, group.by = "cluster_name") + theme(text = element_text(size = 10))
```


```{r, fig.width=12, fig.height=7}
ggarrange(umap_hg19,nh_graph_hg19,beeswarm_hg19, nrow=1, common.legend=F, legend="right")


ggarrange(umap_hg19,nh_graph_resistance_hg19, nrow=1, common.legend=F, legend="right")
```





Now we'll do the same for mouse cells

```{r}
mm10_scRNA_milo <- buildNhoodGraph(mm10_scRNA_milo)

## Plot single-cell UMAP
umap_mm10 <- plotReducedDim(mm10_scRNA_milo, dimred = "UMAP", colour_by="cluster_name", text_by = "cluster_name", text_size = 3) +
  guides(fill="none")

## Plot neighbourhood graph
nh_graph_mm10 <- plotNhoodGraphDA(mm10_scRNA_milo, comparison_subtype, layout="UMAP",alpha=0.05)
nh_graph_resistance_mm10 <- plotNhoodGraphDA(mm10_scRNA_milo, comparison_resistance, layout="UMAP",alpha=0.05)

```

```{r}
comparison_subtype_mm10 <- annotateNhoods(mm10_scRNA_milo, comparison_subtype_mm10, coldata_col = "cluster_name")
head(comparison_subtype_mm10)
beeswarm_mm10 <- plotDAbeeswarm(comparison_subtype_mm10, group.by = "cluster_name") + theme(text = element_text(size = 10)) 

comparison_resistance_mm10 <- annotateNhoods(mm10_scRNA_milo, comparison_resistance_mm10, coldata_col = "cluster_name")
head(comparison_resistance_mm10)
beeswarm_drug_r_mm10 <- plotDAbeeswarm(comparison_resistance_mm10, group.by = "cluster_name") + theme(text = element_text(size = 10))
```

```{r, fig.width=12, fig.height=7}
ggarrange(umap_mm10,nh_graph_mm10,beeswarm_mm10, nrow=1, common.legend=F, legend="right")


ggarrange(umap_mm10,nh_graph_resistance_mm10, nrow=1, common.legend=F, legend="right")
```




# Differential expression between cell populations

We will create a new object because this pseudobulk methodoloy requires raw counts rather than normalized, scaled ones.

## Human cells

```{r}
metadata_hg19 <- hg19_scRNA@meta.data
metadata_hg19$cluster_name <- factor(hg19_scRNA@active.ident)

hg19_scRNA_pseudo <- SingleCellExperiment(assays=list(counts=hg19_scRNA@assays$RNA@counts),
                                        colData=metadata_hg19)


summed_counts_hg19 <- aggregateAcrossCells(hg19_scRNA_pseudo, 
    ids=colData(hg19_scRNA_pseudo)[,c("cluster_name","drug_resistance","subtype")], use.assay.type="counts")
summed_counts_hg19
```


```{r}
# Remove combinations with few cells
summed_counts_filt_hg19 <- summed_counts_hg19[,summed_counts_hg19$ncells>=10]

# We will use this wrapper function from the scatter package, that combines all steps required for differential expression analysis with edgeR
DE_subtype_hg19 <- pseudoBulkDGE(summed_counts_filt_hg19, 
    label=summed_counts_filt_hg19$cluster_name,
    design=~drug_resistance+subtype, # Effect of subtype, accounting for drug resistance
    coef="subtypeTriple_negative",
    condition=summed_counts_filt_hg19$subtype 
)

DE_subtype_hg19_genes <- list()

for (i in 1:length(DE_subtype_hg19)){
  filtered_df <- DE_subtype_hg19[[i]] %>% 
    as.data.frame() %>% 
    #dplyr::filter(abs(logFC)>1 & FDR<0.01) %>% 
    mutate(cluster_name=names(DE_subtype_hg19)[i],
           gene=rownames(.))
  DE_subtype_hg19_genes[[i]] <- filtered_df
}

# Same for drug resistance effect

DE_drug_r_hg19 <- pseudoBulkDGE(summed_counts_filt_hg19, 
    label=summed_counts_filt_hg19$cluster_name,
    design=~subtype + drug_resistance, # Effect drug resistance, accounting for subtype
    coef="drug_resistanceSensitive",
    condition=summed_counts_filt_hg19$drug_resistance
)

DE_drug_r_hg19_genes <- list()

for (i in 1:length(DE_drug_r_hg19)){
  filtered_df <- DE_drug_r_hg19[[i]] %>% 
    as.data.frame() %>% 
    #dplyr::filter(abs(logFC)>0.5 & FDR<0.05) %>% 
    mutate(cluster_name=names(DE_drug_r_hg19)[i],
           gene=rownames(.))
  DE_drug_r_hg19_genes[[i]] <- filtered_df
}
```

```{r, fig.width=12, fig.height=7}
volcano_colors <- c("cornflowerblue", "darkred","grey")
names(volcano_colors) <- c("DOWN", "UP", "NO")


plots_subtype_hg19 <- list()

for (i in 1:length(DE_subtype_hg19_genes)){
  df = DE_subtype_hg19_genes[[i]]
  df[is.na(df)] <- 0
  df$diffexpressed <- "NO"
  df$diffexpressed[df$logFC > 1 & df$FDR < 0.01] <- "UP"
  df$diffexpressed[df$logFC < -1 & df$FDR < 0.01] <- "DOWN"
  p <- ggplot(data=df, aes(x=logFC, y=-log10(FDR), col=diffexpressed)) + 
    geom_point() + 
    theme_minimal() + 
    scale_color_manual(values=volcano_colors) +
    ggtitle(paste0(unique(df$cluster_name))) +
    ylim(0,50)
  plots_subtype_hg19[[i]] <-p
}

legend_volcano <- get_legend(plots_subtype_hg19[[1]])

ggarrange(plotlist=plots_subtype_hg19, legend.grob = legend_volcano, nrow = 1, legend="right")

plots_resistance_hg19 <- list()

for (i in 1:length(DE_drug_r_hg19_genes)){
  df = DE_drug_r_hg19_genes[[i]]
  df[is.na(df)] <- 0
  df$diffexpressed <- "NO"
  df$diffexpressed[df$logFC > 1 & df$FDR < 0.01] <- "UP"
  df$diffexpressed[df$logFC < -1 & df$FDR < 0.01] <- "DOWN"
  p <- ggplot(data=df, aes(x=logFC, y=-log10(FDR), col=diffexpressed)) + 
    geom_point() + 
    theme_minimal() + 
    scale_color_manual(values=volcano_colors) +
    ggtitle(paste0(unique(df$cluster_name))) +
    ylim(0,10)
  plots_resistance_hg19[[i]] <-p
}

ggarrange(plotlist=plots_resistance_hg19, legend.grob = legend_volcano, nrow = 1, legend="right")
```



Let's perform a functional enrichment analysis to see the functions these genes participate in:

### Gene ontology

```{r, fig.width=7, fig.height=15}
plot_GO <- function(df, db){
  df <- df %>% dplyr::filter(abs(logFC)>1 & FDR<0.01)
  # Calculate enrichment
  enrichment <- enrichGO(gene = df$gene,
             OrgDb  = db,
             keyType = "SYMBOL",
             ont  = "BP",
             pAdjustMethod = "BH",
             pvalueCutoff  = 0.01,
             qvalueCutoff  = 0.05)
  enrichment <- clusterProfiler::simplify(enrichment, cutoff=0.7, by="p.adjust", select_fun=min)
  
  GO_ggdata <- enrichment %>%
   as_data_frame() %>%
   arrange(Count)
  GO_ggdata$Description <- factor(GO_ggdata$Description, levels = GO_ggdata$Description)
  print(tail(GO_ggdata, n=20L))
  
ggplot(GO_ggdata, aes(x = Description, y = Count, fill = p.adjust)) +
 geom_bar(stat = "identity") +
 scale_colour_viridis_d(begin=0,end=1) +
 coord_flip() +
 ylab("Number of genes") +
 xlab("GO Terms") +
 theme(axis.text.y = element_text(size=10)) +
  ggtitle(paste0("GO enrichment in", unique(df$cluster_name)))
}

# lapply(DE_drug_r_hg19_genes, function(x){plot_GO(df=x, db="org.Hs.eg.db")}) 
# There is a much smaller number of DE so the enrichment is non significant
lapply(DE_subtype_hg19_genes, function(x){plot_GO(df=x, db="org.Hs.eg.db")})

```



## Murine cells

```{r}
metadata_mm10 <- mm10_scRNA@meta.data
metadata_mm10$cluster_name <- factor(mm10_scRNA@active.ident)

mm10_scRNA_pseudo <- SingleCellExperiment(assays=list(counts=mm10_scRNA@assays$RNA@counts),
                                        colData=metadata_mm10)


summed_counts_mm10 <- aggregateAcrossCells(mm10_scRNA_pseudo, 
    ids=colData(mm10_scRNA_pseudo)[,c("cluster_name","drug_resistance","subtype")], use.assay.type="counts")
summed_counts_mm10
```


```{r}
# Remove combinations with few cells
summed_counts_filt_mm10 <- summed_counts_mm10[,summed_counts_mm10$ncells>=10]

# We will use this wrapper function from the scatter package, that combines all steps required for differential expression analysis with edgeR
DE_subtype_mm10 <- pseudoBulkDGE(summed_counts_filt_mm10, 
    label=summed_counts_filt_mm10$cluster_name,
    design=~drug_resistance+subtype, # Effect of subtype, accounting for drug resistance
    coef="subtypeTriple_negative",
    condition=summed_counts_filt_mm10$subtype 
)

DE_subtype_mm10_genes <- list()

for (i in 1:length(DE_subtype_mm10)){
  filtered_df <- DE_subtype_mm10[[i]] %>% 
    as.data.frame() %>% 
    #dplyr::filter(abs(logFC)>1 & FDR<0.01) %>% 
    mutate(cluster_name=names(DE_subtype_mm10)[i],
           gene=rownames(.))
  DE_subtype_mm10_genes[[i]] <- filtered_df
}

# Same for drug resistance effect

DE_drug_r_mm10 <- pseudoBulkDGE(summed_counts_filt_mm10, 
    label=summed_counts_filt_mm10$cluster_name,
    design=~subtype + drug_resistance, # Effect drug resistance, accounting for subtype
    coef="drug_resistanceSensitive",
    condition=summed_counts_filt_mm10$drug_resistance
)

DE_drug_r_mm10_genes <- list()

for (i in 1:length(DE_drug_r_mm10)){
  filtered_df <- DE_drug_r_mm10[[i]] %>% 
    as.data.frame() %>% 
    #dplyr::filter(abs(logFC)>0.5 & FDR<0.05) %>% 
    mutate(cluster_name=names(DE_drug_r_mm10)[i],
           gene=rownames(.))
  DE_drug_r_mm10_genes[[i]] <- filtered_df
}
```


```{r, fig.width=16, fig.height=7}
plots_subtype_mm10 <- list()

for (i in 1:length(DE_subtype_mm10_genes)){
  df = DE_subtype_mm10_genes[[i]]
  df[is.na(df)] <- 0
  df$diffexpressed <- "NO"
  df$diffexpressed[df$logFC > 1 & df$FDR < 0.01] <- "UP"
  df$diffexpressed[df$logFC < -1 & df$FDR < 0.01] <- "DOWN"
  p <- ggplot(data=df, aes(x=logFC, y=-log10(FDR), col=diffexpressed)) + 
    geom_point() + 
    theme_minimal() + 
    scale_color_manual(values=volcano_colors) +
    ggtitle(paste0(unique(df$cluster_name))) +
    ylim(0,15)
  plots_subtype_mm10[[i]] <-p
}

ggarrange(plotlist=plots_subtype_mm10, legend.grob = legend_volcano, nrow = 1, legend="right")

plots_resistance_mm10 <- list()

for (i in 1:length(DE_drug_r_mm10_genes)){
  df = DE_drug_r_mm10_genes[[i]]
  df[is.na(df)] <- 0
  df$diffexpressed <- "NO"
  df$diffexpressed[df$logFC > 1 & df$FDR < 0.01] <- "UP"
  df$diffexpressed[df$logFC < -1 & df$FDR < 0.01] <- "DOWN"
  p <- ggplot(data=df, aes(x=logFC, y=-log10(FDR), col=diffexpressed)) + 
    geom_point() + 
    theme_minimal() + 
    scale_color_manual(values=volcano_colors) +
    ggtitle(paste0(unique(df$cluster_name))) +
    ylim(0,15)
  plots_resistance_mm10[[i]] <-p
}

ggarrange(plotlist=plots_resistance_mm10, legend.grob = legend_volcano, nrow = 1, legend="right")
```


Let's perform a functional enrichment analysis to see the functions these genes participate in:


### Gene ontology

```{r, eval=F}

lapply(DE_drug_r_mm10_genes, function(x){plot_GO(df=x, db="org.Mm.eg.db")})
lapply(DE_subtype_mm10_genes, function(x){plot_GO(df=x, db="org.Mm.eg.db")})

```

There is no significant enrichment of these genes.

