<html> 
  <head> 
    <title>Characterization of breast cancer at the single cell level using transcriptomic and epigenomic data</title> 
  </head> 
  <body> 
  <h1>
    Characterization of breast cancer at the single cell level using transcriptomic and epigenomic data
  </h1>
    <p> This project is the final thesis of my master's degree in bioinformatics and biostatistics at Universitat Oberta de Catalunya (UOC). </p>
    <p>This is a reanalysis of the wonderful work done by <a href="https://pubmed.ncbi.nlm.nih.gov/31152164/9">Grosselin et al, 2019</a>, where I will try to characterize single cell data at transcriptomic and epigenomic level, to try to identify biologically-relevant characteristics or differences between breast cancer subtypes (luminal vs triple negative) or drug-responsiveness (sensitive vs resistant). </p>

<p>So far the work is organized in different Rmd files: </p>

<ul>
  <li>brca_sc_transcriptomics: Characterization of single cell transcriptomic data (QC, integration, clustering and cell annotation). You can also check the report <a href="brca_transcriptomics.html">here</a> </li>
  <li>brca_sc_epigenomics: Characterization of single cell H3K27me3 data (QC, integration, clustering and cell annotation). Check it <a href="brca_sc_epigenomics.html">here</a></li>
  <li>comparative_analysis_scRNA: Comparison between different models in terms of cell type abundance and expression. Find the report <a href="comparative_transcriptomics.html">here</a></li>
  <li>comparative_analysis_scRNA: comparative_analysis_scChIPseq: Comparison between different models in terms of H3K27me3 profiles. The report is <a href="comparative_epigenomics.html">here</a></li>
</ul>

<p>I hope you find this interesting, I am open to any questions and suggestions! </p>

  </body> 
</html> 
